Add support for Silabs EFR32MG12P device and Thunderboard Sense 2

diegosueiro opened the pull request:
This PR adds the support for Silabs EFR32MG-SLTB004A (Thunderboard Sense 2) board in Zephyr.

It adds the Gecko SDK for EFR32MG12P devices into the ext/hal for Silabs, introduces leuart and i2c drivers for Gecko Devices and a sample application to read Relative Humidity and Temperature from Silabs
SI7021 chip via I2C interface.

This PR depends on #8930 (first 3 commits) and as soon as it gets merged I will update this one to remove the redundant commits.

diegosueiro commented:
Ping @chrta .

chrta commented:
Are the emlib and the other gecko sdk files are still on 5.1.2, but the efm23mg  files are from gecko sdk 5.5.0? Does this work? Or should we first update the gecko sdk to 5.5.0 for all already merged silabs boards?

diegosueiro commented:
@chrta ,

It is working without any issues. If you compare the CMSIS headers between these versions you will see that it is more about adding defines and declarations.

The idea is to update the Gecko SDK in the near future.

dbkinder commented on boards/arm/efr32mg_sltb004a/doc/efr32mg_sltb004a.rst:
```diff
@@ -0,0 +1,202 @@
+.. _efr32mg_sltb004a:
+
+EFR32MG-SLTB004A
+###############
```
The title underline must be at least as long as the title above it, so add one more ``#``

diegosueiro replied:
Addressed.

dbkinder commented on boards/arm/efr32mg_sltb004a/doc/efr32mg_sltb004a.rst:
```diff
@@ -0,0 +1,202 @@
+- One bi-color LED and two push buttons
+- Power enable signals for fine grained power-control
+- On-board SEGGER J-Link debugger for easy programming and debugging, which
+includes a USB virtual COM port
```
this line should be indented to the ``O`` on the previous line

diegosueiro replied:
Addressed.

dbkinder commented on boards/arm/efr32mg_sltb004a/doc/efr32mg_sltb004a.rst:
```diff
@@ -0,0 +1,202 @@
+- On-board SEGGER J-Link debugger for easy programming and debugging, which
+includes a USB virtual COM port
+- Mini Simplicity connector for access to energy profiling and advanced wireless
+network debugging
```
same here, indent the line to the ``M``

diegosueiro replied:
Addressed.

dbkinder commented on boards/arm/efr32mg_sltb004a/doc/efr32mg_sltb004a.rst:
```diff
@@ -0,0 +1,202 @@
+
+The default configuration can be found in the defconfig file:
+
+	``boards/arm/efr32mg_sltb004a/efr32mg_sltb004a_defconfig``
```
delete the blank line and indent:

    The default configuration can be found in the defconfig file:
    ``boards/arm/efr32mg_sltb004a/efr32mg_sltb004a_defconfig``.

diegosueiro replied:
Addressed.

dbkinder reviewed (approved):
+1 for docs, thanks.

diegosueiro commented:
@galak and @MaureenHelm ,

Any feedbacks?

galak reviewed (changes requested):
Number of dts related comments.

diegosueiro commented on boards/arm/efm32wg_stk3800/efm32wg_stk3800.dts:
```diff
@@ -20,5 +24,6 @@
 
 &uart0 {
 	current-speed = <115200>;
+	loc = <1>;
```
Do you mean we need to have a loc (location) property for the tx-pin and rx-pin separately?

galak replied:
Looking at the manual, and maybe loc makes sense.  I was tracing the code and came to soc_gpio_configure/GPIO_PinModeSet.  So I'm confused about what location means for which pins are used.

diegosueiro replied:
The loc is used to instruct which pin (ball) is going to be used. As a consequence, we need to pass which port will be used and the alternate configuration (location). Take a look at the section "6.7 Alternate Functionality Overview" in this document:
https://www.silabs.com/documents/public/data-sheets/efr32fg1-datasheet.pdf

diegosueiro replied:
@galak,

What do you say?

Take a look at the function `uart_gecko_init_pins` in **drivers/serial/uart_gecko.c** file to understand how the "loc" property is used. If you want we can slipt the loc for rx and tx because they can be eventually different.

chrta replied:
For the EFM32WG (and all the older EFM32 i think), there is only one loc per U(S)ART, see e.g. https://www.silabs.com/documents/public/reference-manuals/EFM32WG-RM.pdf chapter 17.5.22 USARTn_ROUTE - I/O Routing Register.

This should be considered (one loc per peripheral vs. one loc per pin in a peripheral), if the driver implementation is changed.

diegosueiro commented on arch/arm/soc/silabs_exx32/efm32wg/dts.fixup:
```diff
@@ -7,9 +7,30 @@
 
-#define CONFIG_UART_GECKO_0_BAUD_RATE	SILABS_EFM32_USART_4000C000_CURRENT_SPEED
-#define CONFIG_UART_GECKO_0_IRQ_PRI	SILABS_EFM32_USART_4000C000_IRQ_0_PRIORITY
+#define USART_0_RX_IRQ			SILABS_GECKO_USART_4000C000_IRQ_0
```
I'm trying to keep the consistency with the define names automatically generated by the alias. I can check if I can do something in the devicetree yaml to have these defines automatically generated without needing a fixup.
Doesn't look nice to me to have in the driver a mixture between defines with and without CONFIG_ prefix.
I even sent an email to the mailing list about cleaning up these fixups and make usage of the aliases. Have you seen it?

diegosueiro replied:
@galak,

I got rid of these fixups by adding the "interrupt-names" for the uarts nodes.

diegosueiro commented on arch/arm/soc/silabs_exx32/efr32fg1p/soc_pinmap.h:
```diff
@@ -18,13 +18,13 @@
 #ifdef CONFIG_USART_GECKO_0
-#if (CONFIG_USART_GECKO_0_GPIO_LOC == 0)
+#if (USART_0_LOC == 0)
 #define PIN_USART0_TXD {gpioPortA, 0, gpioModePushPull, 1}
```
@galak,

The "loc" property is not directly related to the pin number.

galak replied:
Yeah, ignore my comments on loc.  However we should rename to location.

diegosueiro commented:
@galak ,

Comments addressed.
Note that this PR has a dependency on PR #8930.

diegosueiro commented:
@galak,

I'm not sure how to solve the merging conflicts in the Shippable testing environment.

chrta commented:
> @galak,
> 
> I'm not sure how to solve the merging conflicts in the Shippable testing environment.

You have to rebase your branch on the current master. Because the soc files moved to another directory, you have to change a few things (board dts file, create soc dtsi file, change path to drivers is board cmake file etc).

diegosueiro commented:
@chrta,

I already rebased and I'm using the new SoC path. 
The test failures are not related to this.

chrta commented:
@diegosueiro Your SoC commit 10045194958167858bf93db5c74b8405501d32fe is still based on an older master branch, e.g. there is no file dts/arm/silabs/mem.h any more.

diegosueiro commented:
@chrta,

You are right. I missed the 912b4a2 commit.
I'll rebase and push again.

chrta commented:
@diegosueiro Because commit 2eb347b174046037139c024e1ab2f63caf549315 is already merged, please delete `soc/arm/silabs_exx32/efr32mg12p/CMakeLists.txt` and `soc/arm/silabs_exx32/efr32mg12p/soc.c`

chrta commented on drivers/serial/leuart_gecko.c:
```diff
@@ -0,0 +1,385 @@
+	.base = (LEUART_TypeDef *) LEUART_1_BASE_ADDRESS,
+	.clock = cmuClock_LEUART0,
+	.baud_rate = LEUART_1_CURRENT_SPEED,
+	.pin_rx = PIN_LEUART0_RXD,
```
Should PIN_LEUART0_RXD be PIN_LEUART1_RXD (also in the next line)?

diegosueiro replied:
Fixed

diegosueiro commented on boards/arm/efr32mg_sltb004a/efr32mg_sltb004a.yaml:
```diff
@@ -0,0 +1,13 @@
+flash: 1024
+toolchain:
+  - zephyr
+  - gccarmemb
```
done

diegosueiro commented:
@galak,

Anything pending to get this pr merged?

ekarlso commented:
hi guys, any idea on getting this in ?

galak commented:
> hi guys, any idea on getting this in ?

looks like needs some updating due to merge conflicts.  Lets get the UART driver from #8930 merged, and than update this after.

ekarlso commented:
@diegosueiro Can you rebase this ontop of https://github.com/ekarlso/zephyr/tree/thunderboard_sense_2 ?

carlescufi commented:
@diegosueiro thanks for the PR, can you please take a look at the failing CI tests? https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/24567/1/console

diegosueiro commented:
@carlescufi and @galak,

Branch rebased and all checks have passed. ;-)

galak commented on dts/bindings/i2c/silabs,gecko-i2c.yaml:
```diff
@@ -0,0 +1,41 @@
+#
+---
+title: Silabs Gecko I2C Controller
+id: silabs,gecko-leuart
```
remove id: line

galak commented on drivers/i2c/i2c_gecko.c:
```diff
@@ -0,0 +1,255 @@
+static void i2c_gecko_config_func_1(struct device *dev);
+
+static struct i2c_gecko_config i2c_gecko_config_1 = {
+	.base = (I2C_TypeDef *)I2C_1_BASE_ADDRESS,
```
CONFIG_I2C_1_BASE_ADDRESS

galak commented on drivers/i2c/i2c_gecko.c:
```diff
@@ -0,0 +1,255 @@
+	.i2cInit = I2C_INIT_DEFAULT,
+	.pin_sda = PIN_I2C1_SDA,
+	.pin_scl = PIN_I2C1_SCL,
+	.loc = I2C_1_LOCATION,
```
CONFIG_IC2_1_..

galak replied:
CONFIG_IC2_1_..

galak commented on drivers/i2c/i2c_gecko.c:
```diff
@@ -0,0 +1,255 @@
+#ifdef CONFIG_I2C_0
+static void i2c_gecko_config_func_0(struct device *dev);
+
+static struct i2c_gecko_config i2c_gecko_config_0 = {
```
CONFIG_IC2_0_.. for things coming from device tree.

galak commented on dts/bindings/serial/silabs,gecko-leuart.yaml:
```diff
@@ -0,0 +1,36 @@
+---
+title: GECKO LEUART
+id: silabs,gecko-leuart
```
drop id line

galak reviewed (changes requested):
All defines coming from DT should be CONFIG_ prefix, like I did for the UART, and having a mapping in dts_fixup.h to a more general define.

galak commented on dts/bindings/gpio/silabs,efr32mg12-gpio-port.yaml:
```diff
@@ -0,0 +1,32 @@
+---
+title: EFR32MG GPIO
+id: silabs,efr32mg-gpio-port
```
remove id:

galak commented:
Now that the uart driver is merged, an you rebase and fixup the other comments I've made.

diegosueiro commented:
@galak,

Branch rebased, suggestions applied, all checks passed, and basic tests on the target working.

galak commented:
I opened up PR #10728 to merge the base SoC support.  The board code needs a little updating (for moving gpio info into device tree).  Also would like to get a little more review on the uart/i2c drivers.

galak reviewed (changes requested):
Seems like the I2C driver doesn't use interrupts, so we should remove the code associated w/setting up interrupts and the isr handler.

diegosueiro commented on drivers/serial/Kconfig.gecko:
```diff
@@ -52,4 +52,4 @@ config USART_GECKO_3
 	  want to use USART3 device.
 
-endif # UART_GECKO
+endif # UART_GECKO
```
This is showing as outdated. I think this is gone after I rebased it the master instead of the PR that was changing this file.

diegosueiro commented on drivers/i2c/i2c_gecko.c:
```diff
@@ -0,0 +1,255 @@
+}
+
+
+static void i2c_gecko_isr(void *arg)
```
I'll remove that after getting the others reviews.

galak replied:
Ok, I've merged the base SoC support, and add a few minor comments on the board code.

diegosueiro replied:
Done

galak commented on boards/arm/efr32mg_sltb004a/board.h:
```diff
@@ -0,0 +1,49 @@
+
+
+/* Push button PB0 (SW0) */
+#define PB0_GPIO_NAME	CONFIG_GPIO_GECKO_PORTD_NAME
```
Move the button and LED into device tree out of board.h

diegosueiro replied:
Done

pfalcon commented on drivers/serial/leuart_gecko.c:
```diff
@@ -0,0 +1,387 @@
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u32_t flags = LEUART_IntGet(base);
+
+	LEUART_IntClear(base, LEUART_IF_TXC);
```
Why do you clear interrupt here? This function is supposed to be side effect free.

diegosueiro replied:
Removed

pfalcon reviewed (changes requested):
(Github glitches today, let's see if my comments were recorded.)

Some questions/suggestions.

pfalcon commented on drivers/serial/leuart_gecko.c:
```diff
@@ -0,0 +1,387 @@
+static void leuart_gecko_irq_tx_enable(struct device *dev)
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+	u32_t mask = LEUART_IEN_TXBL | LEUART_IEN_TXC;
```
My understanding that this enables both "fifo has empty space" and "transmission complete" interrupts. Just a note that many other drivers don't enable "transmission complete" IRQ, because it's not really much useful.

Again, this is just a note of difference to other drivers. (If someone later would need to investigate any issues and path which led to them.
)

diegosueiro replied:
I've just followed the same pattern for `drivers/serial/uart_gecko.c` driver.

pfalcon reviewed (commented):
Just a note.

pfalcon commented:
> @pfalcon can you review the uart driver.

Well, I tried. If there's something to guaranteedly change, is to add more comments ;-).

Otherwise, it's prompted me to prepare https://github.com/zephyrproject-rtos/zephyr/pull/10761 , because docs there still leave much to be desired (i.e. were partially updated previously).

@diegosueiro : Did you try to run samples/subsys/console/echo with this driver? That would a litmus test whether this driver does everything "good enough" or "not good enough". Please be sure to also test clipboard-pasting of reasonable amount of data into the sample (at least CONFIG_CONSOLE_GETCHAR_BUFSIZE should go without any char loss, but on many boards I tried many more than that can go w/o loss, let's say at least 2x).

diegosueiro committed:
drivers: Introduce Silabs leuart shim serial driver

Adds the Low Energy UART shim serial driver for Silabs Gecko Devices.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro committed:
drivers: Introduce Silabs i2c shim driver

Adds the I2C shim driver for Silabs Gecko Devices.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented on drivers/serial/leuart_gecko.c:
```diff
@@ -0,0 +1,387 @@
+{
+	LEUART_TypeDef *base = DEV_BASE(dev);
+
+	LEUART_Tx(base, c);
```
Added a comment.

diegosueiro commented on drivers/serial/Kconfig.leuart_gecko:
```diff
@@ -0,0 +1,31 @@
+
+if LEUART_GECKO
+
+config LEUART_GECKO_0
```
I'm following the same pattern for `drivers/serial/Kconfig.gecko`

diegosueiro commented:
@pfalcon 

> @diegosueiro : Did you try to run samples/subsys/console/echo with this driver? That would a litmus test whether this driver does everything "good enough" or "not good enough". Please be sure to also test clipboard-pasting of reasonable amount of data into the sample (at least CONFIG_CONSOLE_GETCHAR_BUFSIZE should go without any char loss, but on many boards I tried many more than that can go w/o loss, let's say at least 2x).

I tested as you suggested and it is working fine.

pfalcon commented:
> I tested as you suggested and it is working fine.

Great! Let me approve from my side.

pfalcon reviewed (approved):
+1 for UART driver.

galak commented on boards/arm/efr32mg_sltb004a/board.h:
```diff
@@ -0,0 +1,28 @@
+#endif
+
+/* Environment Sensors Enable Pin */
+#define ENV_SENSE_ENABLE_NAME	CONFIG_GPIO_GECKO_PORTF_NAME
```
where is this used?

diegosueiro replied:
This one I used for testing an i2c sensor in the board to validate the i2c driver.
I'll remove it.

diegosueiro replied:
Done

galak reviewed (changes requested):
do we need board.c and some defines in board.h don't seem to be used.

diegosueiro committed:
boards: Add Silabs EFR32MG-SLTB004A (Thunderboard Sense 2) support

The EFR32™ Mighty Gecko Starter Kit EFR32MG-SLTB004A (a.k.a
Thunderboard Sense 2) contains a MCU from the EFR32MG12P family built
on ARM® Cortex®-M4F processor with low power capabilities.

There is an On-Board J-Link Debugger that presents a virtual COM port
for general purpose application serial data transfer with this
interface and a Mass Storage for firmware flashing.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

diegosueiro commented on boards/arm/efr32mg_sltb004a/board.c:
```diff
@@ -0,0 +1,15 @@
+
+#include <init.h>
+
+static int efr32mg_sltb004a_init(struct device *dev)
```
Done.
